adapters/http-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -945,45 +919,46 @@ protected void setNonEmptyResponsePayload(final HttpServerResponse response, fin
-     * @param responseReady A future to complete once one of the following conditions are met:
-     *              <ul>
-     *              <li>the request did not include a <em>hono-ttd</em> parameter or</li>
-     *              <li>a command has been received and the response ready future has not yet been
-     *              completed or</li>
-     *              <li>the ttd has expired</li>
-     *              </ul>
      * @param uploadMessageSpan The OpenTracing Span used for tracking the processing
      *                       of the request.
      * @return A future indicating the outcome of the operation.
      *         <p>
-     *         The future will be completed with the created message consumer or {@code null}, if
-     *         the response can be sent back to the device without waiting for a command.
+     *         In case of success the future will be completed with a {@link  ResponseReadyTracker} that:
+     *           <ul>
+     *             <li>if device has specified positive ttd will waits for a command:</li>
+     *             <li>otherwise will not wait for a command</li>
+     *           </ul>
+     *         In both cases allocated resources shall be released by the {@link ResponseReadyTracker} itself without any caller interactions.
      *         <p>
      *         The future will be failed with a {@code ServiceInvocationException} if the
      *         message consumer could not be created.
      * @throws NullPointerException if any of the parameters other than TTD or gatewayId is {@code null}.
      */
-    protected final Future<CommandConsumer> createCommandConsumer(
+    protected final Future<ResponseReadyTracker> createResponseReadyTracker(
             final Integer ttdSecs,
             final TenantObject tenantObject,
             final String deviceId,
             final String gatewayId,
             final RoutingContext ctx,
-            final Handler<AsyncResult<Void>> responseReady,
             final Span uploadMessageSpan) {
 
         Objects.requireNonNull(tenantObject);
         Objects.requireNonNull(deviceId);
         Objects.requireNonNull(ctx);
-        Objects.requireNonNull(responseReady);
         Objects.requireNonNull(uploadMessageSpan);
 
-        if (ttdSecs == null || ttdSecs <= 0) {
+        if (ttdSecs == null) {
             // no need to wait for a command
-            responseReady.handle(Future.succeededFuture());
-            return Future.succeededFuture();
+            return ResponseReadyTracker.nop();
         }
+
         final AtomicBoolean requestProcessed = new AtomicBoolean(false);
+
+        // A promise to complete once one of the following conditions are met:
+        // - the request did not include a <em>hono-ttd</em> parameter or
+        // - a command has been received and the response ready future has not yet been completed or<
+        // - the ttd has expired
+        final Promise<Void> responseReady = Promise.promise();
```

Comment:
The `the request did not include a <em>hono-ttd</em> parameter` part doesn't seem to apply here.

I guess one could say that the `responseReady` Promise *has* to be completed at some point, if command consumer creation was successful and the surrounding method returns a succeeded ResponseReadyTracker Future. And that the `responseReady` Promise will be completed either by an incoming command, the ttd getting expired or the HTTP request getting cancelled. Right?

Reply:
Right. Will fix comment